src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1179,22 +1180,41 @@
 (pls/defn-validated deactivate-node!
   "Deactivate the given host, recording the current time. If the node is
   currently inactive, no change is made."
-  [certname :- String]
-  (sql/do-prepared "UPDATE certnames SET deactivated = ?
+  [certname :- String & [timestamp :- pls/Timestamp]]
+  (let [timestamp (to-timestamp (or timestamp (now)))]
+   (sql/do-prepared "UPDATE certnames SET deactivated = ?
                     WHERE certname=? AND deactivated IS NULL"
-                   [(to-timestamp (now)) certname]))
+                    [timestamp certname])))
+
+
+(defn- timestamp-of-newest-record [entity certname time-field]
+  (let [query {:select [time-field]
+               :from [entity]
+               :where [:= :certname certname]
+               :order-by [[time-field :desc]]
+               :limit 1}]
+    (sql/with-query-results result-set (hcore/format query)
+      (time-field (first result-set)))))
+
+(pls/defn-validated have-record-produced-after?
+  [entity :- s/Keyword
+   certname :- String
+   time :- pls/Timestamp]
+  (let [time (to-timestamp time)
+        time-field (if (= entity :reports) :start_time :producer_timestamp)]
```

Comment:
there may be some subtlety here around using start_time vs end_time. If the puppet run takes a very long time for some reason, the two could be quite different. I'm not sure what the right choice is exactly, but I think there's a possibility here that a puppet run starts at t=0, deactivate node issued at t=1, puppet run completes at t=2, report processed at t=3, and node deactivate is processed at t=4. I think my expectation under this situation would be for the deactivate to fail (because a report has come in), but as written here I believe it would would succeed. receive_time would be yet another candidate timestamp. I'm not sure that it's wrong as written, I think I just need a bit more clarity on the intent.